webapp/src/js/services/telemetry.js

```diff
@@ -0,0 +1,222 @@
+    // tell if multiple users use the same device
+    var DEVICE_ID_KEY = 'medic-telemetry-device-id';
+
+    // Intentionally scoped to the specific user, as they may perform a
+    // different role (online vs. offline being being the most obvious) with
+    // different performance implications
+    var DB_ID_KEY = ['medic', Session.userCtx().name, 'telemetry-db'].join('-');
+    var LAST_AGGREGATED_DATE_KEY = [
+      'medic',
+      Session.userCtx().name,
+      'telemetry-date',
+    ].join('-');
+
+    var getDb = function() {
+      var dbName = $window.localStorage.getItem(DB_ID_KEY);
+      if (!dbName) {
+        dbName =
+          'medic-user-' + Session.userCtx().name + '-telemetry-' + uuid();
+        $window.localStorage.setItem(DB_ID_KEY, dbName);
+      }
+      return $window.PouchDB(dbName); // avoid angular-pouch as digest isn't necessary here
+    };
+
+    var getUniqueDeviceId = function() {
+      var uniqueDeviceId = $window.localStorage.getItem(DEVICE_ID_KEY);
+      if (!uniqueDeviceId) {
+        uniqueDeviceId = uuid();
+        $window.localStorage.setItem(DEVICE_ID_KEY, uniqueDeviceId);
+      }
+
+      return uniqueDeviceId;
+    };
+
+    var getLastAggregatedDate = function() {
+      var date = parseInt(
+        $window.localStorage.getItem(LAST_AGGREGATED_DATE_KEY)
+      );
+      if (!date) {
+        date = Date.now();
+        $window.localStorage.setItem(LAST_AGGREGATED_DATE_KEY, date);
+      }
+      return date;
+    };
+
+    var storeIt = function(db, key, value) {
+      return db.post({
+        key: key,
+        value: value,
+        date_recorded: Date.now(),
+      });
+    };
+
+    var submitIfNeeded = function(db) {
+      var monthStart = moment().startOf('month');
+      var dbDate = moment(getLastAggregatedDate());
+      if (dbDate.isBefore(monthStart)) {
+        return aggregate(db).then(function() {
+          return reset(db);
+        });
+      }
+    };
+
+    var convertReduceToKeyValues = function(reduce) {
+      var kv = {};
+      reduce.rows.forEach(function(row) {
+        kv[row.key] = row.value;
+      });
+
+      return kv;
+    };
+
+    var generateAggregateDocId = function(metadata) {
+      return [
+        'telemetry',
+        metadata.year,
+        metadata.month,
+        metadata.user,
+        metadata.deviceId,
+      ].join('-');
+    };
+
+    var generateMetadataSection = function() {
+      var date = moment(getLastAggregatedDate());
+      return {
+        year: date.year(),
+        month: date.month(),
+        user: Session.userCtx().name,
+        deviceId: getUniqueDeviceId(),
+        // REVIEWER: is there anything else we think we should store here?
+        // Candidates would be generic metadata that is not directly device
+        // related or DB related
+      };
+    };
+
+    var generateDeviceStats = function() {
+      return {
+        userAgent: $window.navigator.userAgent,
+        language: $window.navigator.language,
```

Comment:
Language is probably not very useful - I think we ignore it anyway, and resort to the language available in the user-settings doc that we can query from couchdb directly.